_stbt/core.py

```diff
@@ -1633,10 +1633,24 @@ def inner(self):
     return inner
 
 
+def _mark_in_is_visible(fn):
+    @functools.wraps(fn)
+    def inner(self):
+        # This increment isn't threadsafe but I can't imagine using the same
+        # FrameObject instance in multiple threads.
+        self._FrameObject__in_is_visible += 1
+        try:
+            return fn(self)
+        finally:
+            self._FrameObject__in_is_visible -= 1
```

Comment:
This isn't threadsafe.  If thread A is calling `.is_visible` and thread B is calling `.text` then thread B may get a result back other than `None`.  Making this threadsafe while preserving the lazy evaluation of `is_visible` is going to be rather tricky. You could make this counter thread local, or add locking - both of which might affect performance, and would definitely make the code more complex.

I think it would be better to change the semantics of `FrameObject` and make `is_visible` eagerly evaluated.  We can call it in the constructor so we have the guarantee that `is_visible` will be called before any other method or property.  That way we won't need locking, we can just check a simple flag.

Hopefully there aren't too many instances in the wild where we create `FrameObject` instances but don't ever check any of the properties on them.